src/nifreeze/data/dmri.py

```diff
@@ -37,76 +37,118 @@
 from typing_extensions import Self
 
 from nifreeze.data.base import BaseDataset, _cmp, _data_repr
+from nifreeze.data.dmriutils import (
+    DEFAULT_HIGHB_THRESHOLD,
+    DEFAULT_LOWB_THRESHOLD,
+    DEFAULT_MULTISHELL_BIN_COUNT_THR,
+    DEFAULT_NUM_BINS,
+    DTI_MIN_ORIENTATIONS,
+    GRADIENT_BVAL_BVEC_PRIORITY_WARN_MSG,
+    GRADIENT_DATA_MISSING_ERROR,
+    GRADIENT_EXPECTED_COLUMNS_ERROR_MSG,
+    GRADIENT_VOLUME_DIMENSIONALITY_MISMATCH_ERROR,
+    find_shelling_scheme,
+    format_gradients,
+    transform_fsl_bvec,
+)
 from nifreeze.utils.ndimage import get_data, load_api
 
-DEFAULT_CLIP_PERCENTILE = 75
-"""Upper percentile threshold for intensity clipping."""
 
-DEFAULT_MIN_S0 = 1e-5
-"""Minimum value when considering the :math:`S_{0}` DWI signal."""
+def validate_gradients(
+    inst: DWI,
+    attr: attrs.Attribute,
+    value: npt.NDArray[np.floating],
+) -> None:
+    """Strict validator for use in attribute validation (e.g. attrs / validators).
 
-DEFAULT_MAX_S0 = 1.0
-"""Maximum value when considering the :math:`S_{0}` DWI signal."""
+    Ensures row-major convention for gradient table.
 
-DEFAULT_LOWB_THRESHOLD = 50
-"""The lower bound for the b-value so that the orientation is considered a DW volume."""
+    This function is intended for use as an attrs-style validator.
 
-DEFAULT_HIGHB_THRESHOLD = 8000
-"""A b-value cap for DWI data."""
+    Parameters
+    ----------
+    inst : :obj:`~nifreeze.data.dmri.DWI`
+        The instance being validated (unused; present for validator signature).
+    attr : :obj:`~attrs.Attribute`
+        The attribute being validated; attr.name is used in the error message.
+    value : :obj:`~npt.NDArray`
+        The value to validate.
 
-DEFAULT_NUM_BINS = 15
-"""Number of bins to classify b-values."""
+    Raises
+    ------
+    :exc:`ValueError`
+        If the gradient table is invalid.
 
-DEFAULT_MULTISHELL_BIN_COUNT_THR = 7
-"""Default bin count to consider a multishell scheme."""
 
-DTI_MIN_ORIENTATIONS = 6
-"""Minimum number of nonzero b-values in a DWI dataset."""
+    Examples
+    --------
+    Non-finite inputs are rejected::
+
+        >>> validate_gradients(None, None, [[np.inf, 0.0, 0.0, 1000]])
+        Traceback (most recent call last):
+        ...
+        ValueError: Gradient table contains NaN or infinite values.
+        >>> validate_gradients(None, None, [[np.nan, 0.0, 0.0, 1000]])
+        Traceback (most recent call last):
+        ...
+        ValueError: Gradient table contains NaN or infinite values.
+
+    """
+    if np.shape(value)[1] != 4:
+        raise ValueError(GRADIENT_EXPECTED_COLUMNS_ERROR_MSG)
```

Comment:
The formatter will never allow this condition to occur. Adding a doctest to exercise this protection.